drainer/relay.go

```diff
@@ -0,0 +1,156 @@
+package drainer
+
+import (
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb-binlog/drainer/checkpoint"
+	"github.com/pingcap/tidb-binlog/drainer/relay"
+	"github.com/pingcap/tidb-binlog/drainer/sync"
+	"github.com/pingcap/tidb-binlog/pkg/loader"
+	obinlog "github.com/pingcap/tidb-tools/tidb-binlog/slave_binlog_proto/go-binlog"
+	"go.uber.org/zap"
+)
+
+func feedByRelayLogIfNeed(cfg *Config) error {
+	if !cfg.SyncerCfg.Relay.IsEnabled() {
+		return nil
+	}
+
+	// for the mysql type checkpoint
+	// clusterID will be use as the key
+	// we can't get the cluster id from pd so we just set 0
+	// and the checkpoint will use the clusterID exist at the checkpoint table.
+	cpCfg, err := GenCheckPointCfg(cfg, 0 /* clusterID */)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	scfg := cfg.SyncerCfg
+
+	cp, err := checkpoint.NewCheckPoint(cpCfg)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	defer cp.Close()
+
+	if cp.Status() == checkpoint.StatusNormal {
+		return nil
+	}
+
+	reader, err := relay.NewReader(scfg.Relay.LogDir, 1 /* readBufferSize */)
+	if err != nil {
+		return errors.Annotate(err, "failed to create reader")
+	}
+
+	db, ld, err := sync.CreateLoader(scfg.To, scfg.WorkerCount, scfg.TxnBatch,
+		queryHistogramVec, scfg.StrSQLMode, scfg.DestDBType, nil /*loopbacksync.LoopBackSync*/)
+	if err != nil {
+		return errors.Annotate(err, "failed to create loader")
+	}
+
+	defer db.Close()
+
+	err = feedByRelayLog(reader, ld, cp)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// feedByRelayLog will take over the `ld loader.Loader`.
+func feedByRelayLog(r relay.Reader, ld loader.Loader, cp checkpoint.CheckPoint) error {
+	checkpointTS := cp.TS()
+	lastSuccessTS := checkpointTS
+	r.Run()
+
+	loaderQuit := make(chan struct{})
+	var loaderErr error
+	go func() {
+		ld.SetSafeMode(true)
+		loaderErr = ld.Run()
+		close(loaderQuit)
+	}()
+
+	var readerTxnsC <-chan *obinlog.Binlog
+	var toPushLoaderTxn *loader.Txn
+	var loaderInputC chan<- *loader.Txn
+	successTxnC := ld.Successes()
+
+	readerTxnsC = r.Binlogs()
+	readerTxnsCClosed := false
+
+	loaderClosed := false
+
+	for {
+		// when reader is drained and all txn has been push into loader
+		// we close cloader.
+		if readerTxnsC == nil && loaderInputC == nil && !loaderClosed {
+			ld.Close()
+			loaderClosed = true
+		}
+
+		// break once we drainer the success items return by loader.
+		if loaderClosed && successTxnC == nil {
+			break
+		}
+
+		select {
```

Comment:
The logic in this `select` is a bit hard to follow and reason about. Will it be simpler if we just we two separate goroutine for reading and writing?

Reply:
actually first try in multi goroutine style first, but finally, give up, will keep the code style now.